test/cppsuite/test_harness/util/scoped_types.h

```diff
@@ -0,0 +1,165 @@
+ * means.
+ *
+ * In jurisdictions that recognize copyright laws, the author or authors
+ * of this software dedicate any and all copyright interest in the
+ * software to the public domain. We make this dedication for the benefit
+ * of the public at large and to the detriment of our heirs and
+ * successors. We intend this dedication to be an overt act of
+ * relinquishment in perpetuity of all present and future rights to this
+ * software under copyright law.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+namespace test_harness {
+
+class scoped_cursor {
+    public:
+    scoped_cursor() = default;
+    scoped_cursor(WT_SESSION *session, const char *uri, const char *cfg)
+    {
+        reinit(session, uri, cfg);
+    }
+
+    virtual ~scoped_cursor()
+    {
+        if (_cursor != nullptr)
+            testutil_check(_cursor->close(_cursor));
+    }
+
+    /* Moving is ok but copying is not. */
+    scoped_cursor(scoped_cursor &&other)
+    {
+        std::swap(_cursor, other._cursor);
+    }
+
+    scoped_cursor &
+    operator=(scoped_cursor &&other)
+    {
+        scoped_cursor tmp(std::move(other));
+        std::swap(_cursor, tmp._cursor);
+        return (*this);
+    }
+
+    scoped_cursor(const scoped_cursor &) = delete;
+    scoped_cursor &operator=(const scoped_cursor &) = delete;
+
+    void
+    reinit(WT_SESSION *session, const char *uri, const char *cfg)
+    {
+        if (_cursor != nullptr) {
+            testutil_check(_cursor->close(_cursor));
+            _cursor = nullptr;
+        }
+        if (session != nullptr)
+            testutil_check(session->open_cursor(session, uri, nullptr, cfg, &_cursor));
+    }
+
+    WT_CURSOR &
+    operator*()
+    {
+        return (*_cursor);
+    }
+
+    WT_CURSOR *
+    operator->()
+    {
+        return (_cursor);
+    }
+
+    WT_CURSOR *
+    get()
+    {
+        return (_cursor);
+    }
+
+    private:
+    WT_CURSOR *_cursor = nullptr;
+};
+
+class scoped_session {
+    public:
+    scoped_session() = default;
+    explicit scoped_session(WT_CONNECTION *conn)
+    {
+        reinit(conn);
+    }
+
+    virtual ~scoped_session()
+    {
+        if (_session != nullptr)
+            testutil_check(_session->close(_session, nullptr));
+    }
+
+    /* Moving is ok but copying is not. */
+    scoped_session(scoped_session &&other)
+    {
+        std::swap(_session, other._session);
+    }
+
+    scoped_session &
+    operator=(scoped_session &&other)
+    {
+        scoped_session tmp(std::move(other));
+        std::swap(_session, tmp._session);
```

Comment:
Can you add some comments around the more C++y things that you've done here so its clear what is happening to non c++ readers?

Reply:
2190aae